// attio/lib/client/components/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Widget = exports.Typography = exports.TextBlock = exports.Experimental_Table = exports.StatusBadge = exports.Section = exports.Row = exports.LoadingState = exports.Link = exports.Json = exports.Divider = exports.DialogList = exports.DescriptionList = exports.Column = exports.Button = exports.Banner = exports.Badge = exports.Avatar = void 0;
var avatar_js_1 = require("./avatar.js");
Object.defineProperty(exports, "Avatar", { enumerable: true, get: function () { return avatar_js_1.Avatar; } });
var badge_js_1 = require("./badge.js");
Object.defineProperty(exports, "Badge", { enumerable: true, get: function () { return badge_js_1.Badge; } });
var banner_js_1 = require("./banner.js");
Object.defineProperty(exports, "Banner", { enumerable: true, get: function () { return banner_js_1.Banner; } });
var button_js_1 = require("./button.js");
Object.defineProperty(exports, "Button", { enumerable: true, get: function () { return button_js_1.Button; } });
var column_js_1 = require("./column.js");
Object.defineProperty(exports, "Column", { enumerable: true, get: function () { return column_js_1.Column; } });
var description_list_js_1 = require("./description-list.js");
Object.defineProperty(exports, "DescriptionList", { enumerable: true, get: function () { return description_list_js_1.DescriptionList; } });
var dialog_list_js_1 = require("./dialog-list.js");
Object.defineProperty(exports, "DialogList", { enumerable: true, get: function () { return dialog_list_js_1.DialogList; } });
var divider_js_1 = require("./divider.js");
Object.defineProperty(exports, "Divider", { enumerable: true, get: function () { return divider_js_1.Divider; } });
var json_js_1 = require("./json.js");
Object.defineProperty(exports, "Json", { enumerable: true, get: function () { return json_js_1.Json; } });
var link_js_1 = require("./link.js");
Object.defineProperty(exports, "Link", { enumerable: true, get: function () { return link_js_1.Link; } });
var loading_state_js_1 = require("./loading-state.js");
Object.defineProperty(exports, "LoadingState", { enumerable: true, get: function () { return loading_state_js_1.LoadingState; } });
var row_js_1 = require("./row.js");
Object.defineProperty(exports, "Row", { enumerable: true, get: function () { return row_js_1.Row; } });
var section_js_1 = require("./section.js");
Object.defineProperty(exports, "Section", { enumerable: true, get: function () { return section_js_1.Section; } });
var status_badge_js_1 = require("./status-badge.js");
Object.defineProperty(exports, "StatusBadge", { enumerable: true, get: function () { return status_badge_js_1.StatusBadge; } });
var table_js_1 = require("./table.js");
Object.defineProperty(exports, "Experimental_Table", { enumerable: true, get: function () { return table_js_1.Experimental_Table; } });
var text_block_js_1 = require("./text-block.js");
Object.defineProperty(exports, "TextBlock", { enumerable: true, get: function () { return text_block_js_1.TextBlock; } });
var typography_js_1 = require("./typography.js");
Object.defineProperty(exports, "Typography", { enumerable: true, get: function () { return typography_js_1.Typography; } });
var widget_js_1 = require("./widget.js");
Object.defineProperty(exports, "Widget", { enumerable: true, get: function () { return widget_js_1.Widget; } });
