// attio/lib/build/proxy-server-modules-plugin.js
import path from "path";
function getServerModuleProxy(moduleHash) {
    return `
export default async function(...args) {
    return runServerFunction(${JSON.stringify(moduleHash)}, args)
}`;
}
const PLUGIN_NAME = "proxy-server-modules-plugin";
export function proxyServerModulesPlugin({ srcDir }) {
    return {
        name: PLUGIN_NAME,
        setup(build) {
            build.onResolve({ filter: /\.server$/ }, (args) => {
                const filePath = path.relative(srcDir, path.resolve(args.resolveDir, args.path));
                return {
                    path: args.path,
                    namespace: PLUGIN_NAME,
                    pluginData: { moduleHash: filePath },
                };
            });
            build.onLoad({ filter: /.*/, namespace: PLUGIN_NAME }, async (args) => {
                const pluginData = args.pluginData;
                return {
                    contents: getServerModuleProxy(pluginData.moduleHash),
                    loader: "js",
                };
            });
        },
    };
}
