// attio/lint.cjs
module.exports = {
    rules: {
        "attio-client-import": {
            meta: {
                type: "problem",
                docs: {
                    description: 'Disallow imports from "attio/client" in server files',
                    category: "Possible Errors",
                    recommended: true,
                },
                fixable: null,
            },
            create(context) {
                return {
                    // biome-ignore lint/style/useNamingConvention: migration
                    ImportDeclaration(node) {
                        if (node.source.value === "attio/client") {
                            const filename = context.getFilename()
                            if (filename.endsWith(".server.ts")) {
                                context.report({
                                    node,
                                    message:
                                        'Imports from "attio/client" are not allowed in .server.ts files',
                                })
                            } else if (filename.endsWith(".server.js")) {
                                context.report({
                                    node,
                                    message:
                                        'Imports from "attio/client" are not allowed in .server.js files',
                                })
                            }
                        }
                    },
                }
            },
        },
        "server-default-export": {
            meta: {
                type: "problem",
                docs: {
                    description: "Ensure server files have a default exported async function",
                    category: "Possible Errors",
                    recommended: true,
                },
                fixable: null,
            },
            create(context) {
                return {
                    // biome-ignore lint/style/useNamingConvention: migration
                    Program(node) {
                        const filename = context.getFilename()
                        if (filename.endsWith(".server.ts") || filename.endsWith(".server.js")) {
                            const hasDefaultExportAsyncFunction = node.body.some(
                                (statement) =>
                                    statement.type === "ExportDefaultDeclaration" &&
                                    statement.declaration.type === "FunctionDeclaration" &&
                                    statement.declaration.async === true
                            )

                            if (!hasDefaultExportAsyncFunction) {
                                if (filename.endsWith(".server.ts")) {
                                    context.report({
                                        node,
                                        message:
                                            "Server files (.server.ts) must have a default exported async function",
                                    })
                                } else if (filename.endsWith(".server.js")) {
                                    context.report({
                                        node,
                                        message:
                                            "Server files (.server.js) must have a default exported async function",
                                    })
                                }
                            }
                        }
                    },
                }
            },
        },
        "form-submit-button": {
            meta: {
                type: "problem",
                docs: {
                    description:
                        "Ensure <Form/> components have exactly one <SubmitButton/> as a direct child, and <SubmitButton/> components only appear as direct children of <Form/> components",
                    category: "Possible Errors",
                    recommended: true,
                },
                fixable: null,
            },
            create(context) {
                return {
                    // biome-ignore lint/style/useNamingConvention: migration
                    JSXElement(node) {
                        // Check if this is a Form component
                        if (node.openingElement.name.name === "Form") {
                            let submitButtonCount = 0

                            // Check direct children only
                            node.children.forEach((child) => {
                                if (child.type === "JSXElement") {
                                    if (child.openingElement.name.name === "SubmitButton") {
                                        submitButtonCount++
                                    }
                                }
                            })

                            if (submitButtonCount === 0) {
                                context.report({
                                    node,
                                    message:
                                        "<Form/> component must have exactly one <SubmitButton/> as a direct child",
                                })
                            } else if (submitButtonCount > 1) {
                                context.report({
                                    node,
                                    message:
                                        "<Form/> component must not have multiple <SubmitButton/> children",
                                })
                            }
                        }
                        // Check if this is a SubmitButton that's not a direct child of Form
                        else if (node.openingElement.name.name === "SubmitButton") {
                            const parent = node.parent
                            if (
                                !parent ||
                                parent.type !== "JSXElement" ||
                                parent.openingElement.name.name !== "Form"
                            ) {
                                context.report({
                                    node,
                                    message:
                                        "<SubmitButton/> must be a direct child of a <Form/> component",
                                })
                            }
                        }
                    },
                }
            },
        },
        "widget-text-children": {
            meta: {
                type: "problem",
                docs: {
                    description:
                        "Ensure <Widget.TextWidget/> direct children are only <Widget.Title/>, <Badge/>, <Widget.Text.Primary/>, <Widget.Text.Secondary/>, or a custom component.",
                    category: "Possible Errors",
                    recommended: true,
                },
                fixable: null,
            },
            create(context) {
                function isValidWidgetTextChild(child) {
                    if (child.type !== "JSXElement") return true // Ignore non-elements (e.g. text, expressions)
                    const name = child.openingElement.name

                    // Allow <Widget.Title/>
                    if (
                        name.type === "JSXMemberExpression" &&
                        name.object.type === "JSXIdentifier" &&
                        name.object.name === "Widget" &&
                        name.property.type === "JSXIdentifier" &&
                        name.property.name === "Title"
                    ) {
                        return true
                    }

                    // Allow <Badge/>
                    if (name.type === "JSXIdentifier" && name.name === "Badge") {
                        return true
                    }

                    // Allow <Widget.Text.Primary/> and <Widget.Text.Secondary/>
                    if (
                        name.type === "JSXMemberExpression" &&
                        name.object.type === "JSXMemberExpression" &&
                        name.object.object.type === "JSXIdentifier" &&
                        name.object.object.name === "Widget" &&
                        name.object.property.type === "JSXIdentifier" &&
                        name.object.property.name === "Text" &&
                        name.property.type === "JSXIdentifier" &&
                        (name.property.name === "Primary" || name.property.name === "Secondary")
                    ) {
                        return true
                    }

                    // Allow custom components (start with uppercase, not built-in HTML)
                    if (name.type === "JSXIdentifier" && /^[A-Z]/.test(name.name)) {
                        return true
                    }

                    return false
                }
                return {
                    // biome-ignore lint/style/useNamingConvention: migration
                    JSXElement(node) {
                        // Check for <Widget.TextWidget>
                        const name = node.openingElement.name
                        if (
                            (name.type === "JSXMemberExpression" &&
                                name.object.type === "JSXIdentifier" &&
                                name.object.name === "Widget" &&
                                name.property.type === "JSXIdentifier" &&
                                name.property.name === "TextWidget") ||
                            (name.type === "JSXIdentifier" && name.name === "Widget.TextWidget")
                        ) {
                            node.children.forEach((child) => {
                                if (!isValidWidgetTextChild(child)) {
                                    context.report({
                                        node: child,
                                        message:
                                            "Direct children of <Widget.TextWidget/> must be <Widget.Title/>, <Badge/>, <Widget.Text.Primary/>, <Widget.Text.Secondary/>, or a custom component.",
                                    })
                                }
                            })
                        }
                    },
                }
            },
        },
    },
}
