// attio/lib/commands/dev/graphql-code-gen.js
import { readFileSync } from "fs";
import chokidar from "chokidar";
import { parse } from "graphql";
import { complete, errored, isErrored } from "@attio/fetchable-npm";
import { generateOperations } from "../../graphql/generate-operations.js";
import { GraphQLError } from "../../graphql/graphql-error.js";
import { findNodeModulesPath } from "../../util/find-node-modules-path.js";
import { hardExit } from "../../util/hard-exit.js";
export async function graphqlCodeGen() {
    try {
        const schemaPath = await findNodeModulesPath(["schema.graphql"]);
        if (!schemaPath) {
            hardExit("No schema.graphql found in node_modules. Did you run `npm install`?");
        }
        const schema = parse(readFileSync(schemaPath, "utf8"));
        await generateOperations(".", schema);
        return complete(true);
    }
    catch (error) {
        if (error instanceof GraphQLError) {
            return errored({ code: "GRAPHQL_ERROR", error: error });
        }
        else {
            return errored({
                code: "GRAPHQL_CODEGEN_ERROR",
                error: error instanceof Error ? error.message : "GraphQL code generation failed",
            });
        }
    }
}
export function watchGraphqlCodegen(onSuccess, onError) {
    const watcher = chokidar.watch(["**/*.graphql", "**/*.gql"], {
        ignored: ["**/node_modules/**", "**/dist/**"],
        cwd: ".",
    });
    watcher.on("ready", async () => {
        const result = await graphqlCodeGen();
        if (isErrored(result)) {
            if (result.error.code === "GRAPHQL_ERROR") {
                onError?.(result.error.error.toString());
            }
            else {
                hardExit(result.error.error.toString());
            }
        }
        else {
            onSuccess?.();
        }
        watcher.on("all", async (event, path) => {
            if (event === "add" || event === "change" || event === "unlink") {
                const result = await graphqlCodeGen();
                if (isErrored(result)) {
                    if (result.error.code === "GRAPHQL_ERROR") {
                        onError?.(result.error.error.toString());
                    }
                    else {
                        hardExit(result.error.error.toString());
                    }
                }
                else {
                    onSuccess?.();
                }
            }
        });
    });
    return () => {
        watcher.close();
    };
}
