// attio/lib/commands/dev/prepare-build-context.js
import path from "path";
import chalk from "chalk";
import { combine, combineAsync, complete, isErrored, map } from "@attio/fetchable-npm";
import { findWorkflowBlockModules } from "../../build/server/find-workflow-block-server-modules.js";
import { ClientBuilder } from "./client-builder.js";
import { ServerBuilder } from "./server-builder.js";
export async function prepareBuildContext(mode) {
    const appDir = "";
    const srcDir = "src";
    const assetsDir = path.join(srcDir, "assets");
    const webhooksDir = path.join(srcDir, "webhooks");
    const eventsDir = path.join(srcDir, "events");
    const buildersResult = await combineAsync({
        client: ClientBuilder.create({
            outfile: path.resolve("dist", "index.js"),
            directories: {
                app: appDir,
                src: srcDir,
                assets: assetsDir,
            },
            mode,
        }),
        server: ServerBuilder.create({
            outfile: path.resolve("dist", "server.js"),
            directories: {
                app: appDir,
                src: srcDir,
                webhooks: webhooksDir,
                events: eventsDir,
            },
            mode,
        }),
    });
    if (isErrored(buildersResult)) {
        return buildersResult;
    }
    const { client, server } = buildersResult.value;
    return complete({
        rebuild: async () => {
            const workflowBlockModulesResult = await findWorkflowBlockModules(srcDir);
            if (isErrored(workflowBlockModulesResult)) {
                return workflowBlockModulesResult;
            }
            const workflowBlockModules = workflowBlockModulesResult.value;
            return combineAsync({
                client: client.rebuild({ workflowBlockModules }),
                server: server.rebuild({ workflowBlockModules }),
            });
        },
        dispose: async (timeoutMs) => {
            const clientResult = await client.dispose(timeoutMs);
            const serverResult = await server.dispose(timeoutMs);
            return map(combine([clientResult, serverResult]), () => undefined);
        },
    });
}
export function printBuildContextError(error) {
    switch (error.code) {
        case "FAILED_TO_CREATE_TEMP_FILE":
            process.stderr.write(`${chalk.red("✖ ")}Failed to create temp file: ${error.path}\n`);
            break;
        case "FAILED_TO_GENERATE_CLIENT_ENTRY":
            process.stderr.write(`${chalk.red("✖ ")}Failed to generate client entry\n`);
            break;
        case "FAILED_TO_DISPOSE_OF_BUILD_CONTEXT":
            process.stderr.write(`${chalk.red("✖ ")}Failed to dispose of build context: ${error.error}\n`);
            break;
        case "FAILED_TO_CREATE_ESBUILD_CONTEXT":
            process.stderr.write(`${chalk.red("✖ ")}Failed to create esbuild context (${error.outfile}): ${error.error}\n`);
            break;
        case "ERROR_FINDING_SURFACE_EXPORTS":
            process.stderr.write(`${chalk.red("✖ ")}Failed to find surface exports: ${error.error}\n`);
            break;
        case "APP_ENTRY_POINT_NOT_FOUND":
            process.stderr.write(`${chalk.red("✖ ")}Could not find app.ts. See https://docs.attio.com/sdk/guides/app-ts-migration\n`);
            break;
        case "APP_EXPORT_NOT_FOUND":
            process.stderr.write(`${chalk.red("✖ ")}Could not find a named export "app" in ${path.basename(error.path)}. See https://docs.attio.com/sdk/guides/app-ts-migration\n`);
            break;
        case "UNPARSABLE_BUILD_ERROR":
            process.stderr.write(`${chalk.red("✖ ")}Failed to parse build error: ${error.error}\n`);
            break;
        case "ERROR_FINDING_SERVER_FUNCTION_MODULES":
            process.stderr.write(`${chalk.red("✖ ")}Failed to find server modules: ${error.cause}\n`);
            break;
        case "WORKFLOW_BLOCK_RESOLUTION_FAILED":
            process.stderr.write(`${chalk.red("✖ ")}Failed to build workflow block(s): ${error.message}\n`);
            break;
        default:
            return error;
    }
}
