// attio/lib/commands/dev/validate-typescript.js
import chokidar from "chokidar";
import { getDiagnostics, readConfig, typeScriptErrorSchema } from "../../util/typescript.js";
import path from "path";
export function validateTypeScript(onSuccess, onError) {
    let isShuttingDown = false;
    const watcher = chokidar.watch(["src/**/*.ts", "src/**/*.tsx"], {
        ignored: [
            "**/node_modules/**",
            "**/dist/**",
            "**/*.graphql.d.ts",
            "**/*.gql.d.ts",
        ],
        cwd: ".",
        persistent: true,
        awaitWriteFinish: {
            stabilityThreshold: 100,
            pollInterval: 100,
        },
    });
    async function handleValidation() {
        if (isShuttingDown)
            return;
        try {
            const program = await readConfig(path.resolve("tsconfig.json"));
            if (program === "Not a TypeScript project") {
                onSuccess?.();
                return;
            }
            const errors = await getDiagnostics(program);
            if (errors.length) {
                onError?.(errors);
            }
            else {
                onSuccess?.();
            }
        }
        catch (error) {
            if (typeof error === "object" &&
                error !== null &&
                "code" in error &&
                (error.code === "ENOENT" || error.code === "EACCES" || error.code === "EPERM")) {
                throw error;
            }
            if (error instanceof Error) {
                const tsError = typeScriptErrorSchema.parse({ text: error.message });
                onError?.([tsError]);
            }
        }
    }
    let watcherReady = false;
    watcher.on("ready", () => {
        watcherReady = true;
        watcher.on("all", (event, path) => {
            if (event === "add" || event === "change" || event === "unlink") {
                handleValidation();
            }
        });
    });
    watcher.on("error", (error) => {
        process.stderr.write(`TypeScript watcher error: ${error}\n`);
    });
    return [
        () => {
            isShuttingDown = true;
            if (watcherReady) {
                try {
                    watcher.close();
                }
                catch (error) {
                    process.stderr.write(`Error closing TypeScript watcher: ${error}\n`);
                }
            }
        },
        handleValidation,
    ];
}
