// attio/lib/util/load-attio-cli-version.js
import { readFileSync } from "fs";
import { fileURLToPath } from "url";
import { findUpSync } from "find-up-simple";
import { z } from "zod";
import { complete, errored } from "@attio/fetchable-npm";
const FILE_NAME = "package.json";
const packageJsonSchema = z.object({
    name: z.literal("attio"),
    version: z.string({
        required_error: "No CLI version found",
        invalid_type_error: "CLI version must be a string in package.json",
    }),
});
let packageJson;
export function loadAttioCliVersion() {
    if (packageJson === undefined) {
        const cwd = fileURLToPath(import.meta.url);
        const packageJsonPath = findUpSync(FILE_NAME, { cwd });
        if (packageJsonPath === undefined) {
            return errored({
                code: "UNABLE_TO_FIND_PACKAGE_JSON",
                path: cwd,
            });
        }
        let contents;
        try {
            contents = readFileSync(packageJsonPath, "utf8");
        }
        catch (error) {
            return errored({
                code: "UNABLE_TO_READ_PACKAGE_JSON",
                error,
            });
        }
        let json;
        try {
            json = JSON.parse(contents);
        }
        catch (error) {
            return errored({
                code: "UNABLE_TO_PARSE_PACKAGE_JSON",
                error,
            });
        }
        const result = packageJsonSchema.safeParse(json);
        if (!result.success) {
            return errored({
                code: "INVALID_PACKAGE_JSON",
                error: result.error,
            });
        }
        packageJson = result.data;
    }
    const { version } = packageJson;
    if (!version) {
        return errored({
            code: "NO_CLI_VERSION_FOUND",
        });
    }
    return complete(version);
}
