// attio/lib/util/spinner.js
import chalk from "chalk";
import { isComplete } from "@attio/fetchable-npm";
const frames = ["⠋", "⠙", "⠹", "⠸", "⠼", "⠴", "⠦", "⠧", "⠇", "⠏"];
class Spinner {
    frameIndex = 0;
    stopped = false;
    interval = null;
    message = "";
    start(message) {
        if (this.interval) {
            clearInterval(this.interval);
        }
        this.message = message;
        this.stopped = false;
        this.interval = setInterval(() => {
            if (this.stopped)
                return;
            const frame = frames[(this.frameIndex = ++this.frameIndex % frames.length)];
            process.stdout.write(`\r${frame} ${this.message}`);
        }, 80);
        return this;
    }
    update(message) {
        this.message = message;
        return this;
    }
    stop(message) {
        this.stopped = true;
        if (this.interval) {
            clearInterval(this.interval);
            this.interval = null;
        }
        if (message) {
            process.stdout.write(`\r${message}    \n`);
        }
        return this;
    }
    success(message) {
        return this.stop(`${chalk.green("✓")} ${message}`);
    }
    error(message) {
        return this.stop(`${chalk.red("✖")} ${message}`);
    }
    warning(message) {
        return this.stop(`${chalk.yellow("⚠")} ${message}`);
    }
}
export async function spinnerify(busyMessage, successMessage, fn) {
    const spinner = new Spinner();
    spinner.start(busyMessage);
    try {
        const result = await fn();
        if (isComplete(result)) {
            spinner.success(typeof successMessage === "string" ? successMessage : successMessage(result.value));
        }
        return result;
    }
    finally {
        spinner.stop();
    }
}
