// attio/lib/client/workflow/outcome-data-schema/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.timestamp = exports.struct = exports.string = exports.number = exports.boolean = exports.attioSequence = exports.attioSelect = exports.attioRecord = exports.attioObject = exports.attioList = exports.attioAttribute = exports.attioActor = exports.array = void 0;
var array_node_1 = require("./array-node");
Object.defineProperty(exports, "array", { enumerable: true, get: function () { return array_node_1.array; } });
var attio_actor_1 = require("./attio-actor");
Object.defineProperty(exports, "attioActor", { enumerable: true, get: function () { return attio_actor_1.attioActor; } });
var attio_attribute_1 = require("./attio-attribute");
Object.defineProperty(exports, "attioAttribute", { enumerable: true, get: function () { return attio_attribute_1.attioAttribute; } });
var attio_list_1 = require("./attio-list");
Object.defineProperty(exports, "attioList", { enumerable: true, get: function () { return attio_list_1.attioList; } });
var attio_object_1 = require("./attio-object");
Object.defineProperty(exports, "attioObject", { enumerable: true, get: function () { return attio_object_1.attioObject; } });
var attio_record_1 = require("./attio-record");
Object.defineProperty(exports, "attioRecord", { enumerable: true, get: function () { return attio_record_1.attioRecord; } });
var attio_select_1 = require("./attio-select");
Object.defineProperty(exports, "attioSelect", { enumerable: true, get: function () { return attio_select_1.attioSelect; } });
var attio_sequence_1 = require("./attio-sequence");
Object.defineProperty(exports, "attioSequence", { enumerable: true, get: function () { return attio_sequence_1.attioSequence; } });
var boolean_node_1 = require("./boolean-node");
Object.defineProperty(exports, "boolean", { enumerable: true, get: function () { return boolean_node_1.boolean; } });
var number_node_1 = require("./number-node");
Object.defineProperty(exports, "number", { enumerable: true, get: function () { return number_node_1.number; } });
var string_node_1 = require("./string-node");
Object.defineProperty(exports, "string", { enumerable: true, get: function () { return string_node_1.string; } });
var struct_node_1 = require("./struct-node");
Object.defineProperty(exports, "struct", { enumerable: true, get: function () { return struct_node_1.struct; } });
var timestamp_node_1 = require("./timestamp-node");
Object.defineProperty(exports, "timestamp", { enumerable: true, get: function () { return timestamp_node_1.timestamp; } });
