// attio/lib/client/index.js
"use strict";
/**
 * This module defines the types of the "attio/client" SDK.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.experimental_Workflow = exports.showToast = exports.showIframe = exports.showDialog = exports.runQuery = exports.platform = exports.Forms = exports.confirm = exports.alert = void 0;
const tslib_1 = require("tslib");
var alert_js_1 = require("./alert.js");
Object.defineProperty(exports, "alert", { enumerable: true, get: function () { return alert_js_1.alert; } });
tslib_1.__exportStar(require("./components/index.js"), exports);
var confirm_js_1 = require("./confirm.js");
Object.defineProperty(exports, "confirm", { enumerable: true, get: function () { return confirm_js_1.confirm; } });
tslib_1.__exportStar(require("./experimental/index.js"), exports);
exports.Forms = tslib_1.__importStar(require("./forms/index.js"));
tslib_1.__exportStar(require("./forms/path.js"), exports);
tslib_1.__exportStar(require("./hooks/index.js"), exports);
var platform_js_1 = require("./platform.js");
Object.defineProperty(exports, "platform", { enumerable: true, get: function () { return platform_js_1.platform; } });
var run_query_js_1 = require("./run-query.js");
Object.defineProperty(exports, "runQuery", { enumerable: true, get: function () { return run_query_js_1.runQuery; } });
var show_dialog_js_1 = require("./show-dialog.js");
Object.defineProperty(exports, "showDialog", { enumerable: true, get: function () { return show_dialog_js_1.showDialog; } });
var show_iframe_js_1 = require("./show-iframe.js");
Object.defineProperty(exports, "showIframe", { enumerable: true, get: function () { return show_iframe_js_1.showIframe; } });
var show_toast_js_1 = require("./show-toast.js");
Object.defineProperty(exports, "showToast", { enumerable: true, get: function () { return show_toast_js_1.showToast; } });
// biome-ignore lint/style/useNamingConvention: Experimental
exports.experimental_Workflow = tslib_1.__importStar(require("./workflow/index.js"));
