// attio/lib/build/client/create-client-build-config.js
import globalExternals from "@fal-works/esbuild-plugin-global-externals";
import { proxyBlockModulesPlugin } from "../proxy-block-modules-plugin.js";
import { proxyServerModulesPlugin } from "../proxy-server-modules-plugin.js";
export function createClientBuildConfig({ appDir, srcDir, entryPoint, workflowBlockModulesRef, }) {
    return {
        entryPoints: [entryPoint],
        logLevel: "silent",
        bundle: true,
        platform: "browser",
        format: "iife",
        plugins: [
            globalExternals({
                react: {
                    varName: "React",
                    type: "cjs",
                    namedExports: [
                        "default",
                        "Children",
                        "Component",
                        "Fragment",
                        "Profiler",
                        "PureComponent",
                        "StrictMode",
                        "Suspense",
                        "SuspenseList",
                        "cloneElement",
                        "createContext",
                        "createElement",
                        "createFactory",
                        "createRef",
                        "forwardRef",
                        "isValidElement",
                        "lazy",
                        "memo",
                        "startTransition",
                        "unstable_SuspenseList",
                        "useCallback",
                        "useContext",
                        "useDebugValue",
                        "useDeferredValue",
                        "useEffect",
                        "useId",
                        "useImperativeHandle",
                        "useInsertionEffect",
                        "useLayoutEffect",
                        "useMemo",
                        "useReducer",
                        "useRef",
                        "useState",
                        "useSyncExternalStore",
                        "useTransition",
                    ],
                },
                attio: {
                    varName: "ATTIO_ROOT_SDK",
                    type: "cjs",
                },
                "attio/client": {
                    varName: "ATTIO_CLIENT_EXTENSION_SDK",
                    type: "cjs",
                },
            }),
            proxyServerModulesPlugin({ srcDir }),
            proxyBlockModulesPlugin({ appDir, workflowBlockModulesRef }),
        ],
    };
}
