// attio/lib/build/server/fine-workflow-block.test.js
import { describe, expect, it } from "vitest";
import { findWorkflowBlockModulesFromSource } from "./find-workflow-block-server-modules.js";
describe("findWorkflowBlockModulesFromSource", () => {
    it("should find workflow block modules in source code", () => {
        const appTsSource = `
import { block } from "./test-block"
export const app = {
    workflow: {
        blocks:  {
            steps: [block],
        },
    },
}`;
        const blockTsSource = `
import {execute, schema} from "./test-block.block"
export const block = {
    id: "test-block",
    title: "test",
    schema,
    execute,
    Configurator: () => {},
}`;
        const blockBlockTsSource = `
import {experimental_Workflow} from "attio/server"
export const schema = async () => ({
    test: experimental_Workflow.number(),
})
export const execute = async (config) => {}`;
        const getModuleSource = (path) => {
            if (path === "/test-block.ts") {
                return blockTsSource;
            }
            if (path === "/test-block.block.ts") {
                return blockBlockTsSource;
            }
            return "";
        };
        const doesModuleExist = (path) => {
            if (path === "/test-block.ts" || path === "/test-block.block.ts") {
                return true;
            }
            return false;
        };
        const modules = new Map();
        const result = findWorkflowBlockModulesFromSource(appTsSource, "", {
            modules,
            getModuleSource,
            doesModuleExist,
        });
        expect(result.state).toBe("complete");
        if (result.state === "complete") {
            const value = result.value;
            const entries = value.entries();
            expect(value.size).toBe(1);
            expect(entries.next().value).toEqual([
                "test-block",
                {
                    activate: undefined,
                    deactivate: undefined,
                    execute: { export: "execute", path: "/test-block.block.ts" },
                    schema: { export: "schema", path: "/test-block.block.ts" },
                },
            ]);
        }
    });
});
