// attio/lib/build/server/generate-server-entry.js
import fs from "fs/promises";
import path from "path";
import { glob } from "glob";
import { combineAsync, complete, errored, fromPromise, isComplete, isErrored, } from "@attio/fetchable-npm";
import { APP_SETTINGS_FILENAME } from "../../constants/settings-files.js";
async function findServerFunctionModules(cwd, pattern) {
    try {
        return complete(await glob(`**/*.${pattern}.{js,ts}`, { nodir: true, cwd }));
    }
    catch (error) {
        return errored({
            code: "ERROR_FINDING_SERVER_FUNCTION_MODULES",
            path: cwd,
            pattern,
            cause: error,
        });
    }
}
export async function generateServerEntry({ appDirAbsolute, srcDirAbsolute, webhooksDirAbsolute, eventDirAbsolute, workflowBlockModules, log, }) {
    const pathsResult = await combineAsync({
        serverFunctionConcretePaths: findServerFunctionModules(srcDirAbsolute, "server"),
        webhookConcretePaths: findServerFunctionModules(webhooksDirAbsolute, "webhook"),
        eventConcretePaths: findServerFunctionModules(eventDirAbsolute, "event"),
    });
    if (isErrored(pathsResult)) {
        return pathsResult;
    }
    const { serverFunctionConcretePaths, webhookConcretePaths, eventConcretePaths } = pathsResult.value;
    const serverFunctionModules = serverFunctionConcretePaths.map((path) => ({
        path,
        hash: path.replace(/\.(js|ts)$/, ""),
    }));
    const webhookModules = webhookConcretePaths.map((path) => ({
        path,
        id: path.replace(/\.webhook\.(js|ts)$/, ""),
    }));
    const eventModules = eventConcretePaths.map((path) => ({
        path,
        id: path.replace(/\.event\.(js|ts)$/, ""),
    }));
    for (const module of serverFunctionModules) {
        log?.(`🔎 Found server module "${module.path}"`);
    }
    const settingsSchemaResult = await fromPromise(fs.readFile(path.join(srcDirAbsolute, APP_SETTINGS_FILENAME)));
    const initSettingsJS = isComplete(settingsSchemaResult)
        ? `
        import appSettingsSchema from ${JSON.stringify(path.join(srcDirAbsolute, APP_SETTINGS_FILENAME))}

        registerSettingsSchema(appSettingsSchema)
    `
        : `
        registerSettingsSchema({workspace: {}})
    `;
    return complete(`
        ${initSettingsJS}

        const modules = new Map()
        const webhookModules = new Map()
        const eventModules = new Map()
        const workflowModules = new Map()
    
        ${serverFunctionModules
        .map((module) => `modules.set("${module.hash}", () => import(${JSON.stringify(path.join(srcDirAbsolute, module.path))}))`)
        .join("\n")}

        ${webhookModules
        .map((module) => `webhookModules.set("${module.id}", () => import(${JSON.stringify(path.join(webhooksDirAbsolute, module.path))}))`)
        .join("\n")}

        ${eventModules
        .map((module) => `eventModules.set("${module.id}", () => import(${JSON.stringify(path.join(eventDirAbsolute, module.path))}))`)
        .join("\n")}

        ${[...workflowBlockModules.entries()]
        .map(([blockId, handlers]) => `workflowModules.set(
                        ${JSON.stringify(blockId)},
                        {
                        ${[...Object.entries(handlers)]
        .map(([handler, data]) => data
        ? `${JSON.stringify(handler)}: {
                                        module: () => import(${JSON.stringify(path.join(appDirAbsolute, data.path))}),
                                        export: ${JSON.stringify(data.export)}
                                    },`
        : "")
        .join("\n")}
                        }
                    )`)
        .join("\n")}

        var stdin_default;
        var stdin_webhooks_default;
        var stdin_events_default;
        var stdin_workflow_block_handlers_default;

        function main() {
            stdin_default = async function(moduleHash, args) {
                const module = modules.get(moduleHash)

                if (!module) {
                    throw new Error(\`Module \${moduleHash} not found\`)
                }

                const func = (await module()).default

                if (!func) {
                    throw new Error(\`Default export not found in module \${moduleHash}\`)
                }

                if (typeof func !== "function") {
                    throw new Error(\`\${moduleHash} does not export a function\`)
                }

                return func(...args)
            }

            stdin_webhooks_default = async function(webhookModuleId, args) {
                const module = webhookModules.get(webhookModuleId)

                if (!module) {
                    throw new Error(\`Webhook handler not found: \${webhookModuleId}\`)
                }

                const func = (await module()).default

                if (!func) {
                    throw new Error(\`Default export not found in module \${webhookModuleId}\`)
                }

                if (typeof func !== "function") {
                    throw new Error(\`\${webhookModuleId} does not export a function\`)
                }

                return func(...args)
            }

            stdin_events_default = async function(eventModuleId, args) {
                const module = eventModules.get(eventModuleId)

                if (!module) {
                    throw new Error("Event handler not found")
                }

                const func = (await module()).default

                if (!func) {
                    throw new Error(\`Default export not found in module \${eventModuleId}\`)
                }

                if (typeof func !== "function") {
                    throw new Error(\`\${eventModuleId} does not export a function\`)
                }

                return func(...args)
            }

            stdin_workflow_block_handlers_default = async function(blockId, handler, args) {
                const modules = workflowModules.get(blockId)

                if (!modules) {
                    throw new Error("Workflow block modules not found")
                }

                const handlerData = modules[handler]

                if (!handlerData) {
                    throw new Error(\`Workflow block module \${JSON.stringify(handler)} not found\`)
                }

                if (handler === "schema") {
                    const schema = (await handlerData.module())[handlerData.export]

                    if (!schema) {
                        throw new Error(\`"\${handler}\" export not found in module \${JSON.stringify(blockId)}\`)
                    }

                    return schema
                }

                const func = (await handlerData.module())[handlerData.export]

                if (!func) {
                    throw new Error(\`"\${handler}\" export not found in module \${JSON.stringify(blockId)}\`)
                }

                if (typeof func !== "function") {
                   throw new Error(\`"\${handler}\" export in module \${JSON.stringify(blockId)} is not a function\`)
                }

                return args ? func(...args) : func()
            }
        }

        main()

`);
}
