// @hono/graphql-server/dist/index.js
// src/index.ts
import {
  Source,
  parse,
  execute,
  validateSchema,
  validate,
  specifiedRules,
  getOperationAST,
  GraphQLError
} from "graphql";

// src/parse-body.ts
async function parseBody(req) {
  const contentType = req.headers.get("content-type");
  switch (contentType) {
    case "application/graphql":
      return { query: await req.text() };
    case "application/json":
      try {
        return await req.json();
      } catch (e) {
        if (e instanceof Error) {
          console.error(`${e.stack || e.message}`);
        }
        throw Error(`POST body sent invalid JSON: ${e}`);
      }
    case "application/x-www-form-urlencoded":
      return parseFormURL(req);
  }
  return {};
}
var parseFormURL = async (req) => {
  const text = await req.text();
  const searchParams = new URLSearchParams(text);
  const res = {};
  searchParams.forEach((v, k) => res[k] = v);
  return res;
};

// src/index.ts
var graphqlServer = (options) => {
  const schema = options.schema;
  const pretty = options.pretty ?? false;
  const validationRules = options.validationRules ?? [];
  const showGraphiQL = options.graphiql ?? false;
  return async (c) => {
    if (c.req.method !== "GET" && c.req.method !== "POST") {
      return c.json(errorMessages(["GraphQL only supports GET and POST requests."]), 405, {
        Allow: "GET, POST"
      });
    }
    let params;
    try {
      params = await getGraphQLParams(c.req.raw);
    } catch (e) {
      if (e instanceof Error) {
        console.error(`${e.stack || e.message}`);
        return c.json(errorMessages([e.message], [e]), 400);
      }
      throw e;
    }
    const { query, variables, operationName } = params;
    if (query == null) {
      if (showGraphiQL && c.req.method === "GET") {
        return respondWithGraphiQL(c);
      }
      return c.json(errorMessages(["Must provide query string."]), 400);
    }
    const schemaValidationErrors = validateSchema(schema);
    if (schemaValidationErrors.length > 0) {
      return c.json(
        errorMessages(["GraphQL schema validation error."], schemaValidationErrors),
        500
      );
    }
    let documentAST;
    try {
      documentAST = parse(new Source(query, "GraphQL request"));
    } catch (syntaxError) {
      if (syntaxError instanceof Error) {
        console.error(`${syntaxError.stack || syntaxError.message}`);
        const e = new GraphQLError(syntaxError.message, {
          originalError: syntaxError
        });
        return c.json(errorMessages(["GraphQL syntax error."], [e]), 400);
      }
      throw syntaxError;
    }
    const validationErrors = validate(schema, documentAST, [...specifiedRules, ...validationRules]);
    if (validationErrors.length > 0) {
      return c.json(errorMessages(["GraphQL validation error."], validationErrors), 400);
    }
    if (c.req.method === "GET") {
      const operationAST = getOperationAST(documentAST, operationName);
      if (operationAST && operationAST.operation !== "query") {
        return c.json(
          errorMessages([
            `Can only perform a ${operationAST.operation} operation from a POST request.`
          ]),
          405,
          { Allow: "POST" }
        );
      }
    }
    let result;
    const { rootResolver } = options;
    try {
      result = await execute({
        schema,
        document: documentAST,
        rootValue: rootResolver ? await rootResolver(c) : null,
        contextValue: c,
        variableValues: variables,
        operationName
      });
    } catch (contextError) {
      if (contextError instanceof Error) {
        console.error(`${contextError.stack || contextError.message}`);
        const e = new GraphQLError(contextError.message, {
          originalError: contextError,
          nodes: documentAST
        });
        return c.json(errorMessages(["GraphQL execution context error."], [e]), 400);
      }
      throw contextError;
    }
    if (!result.data) {
      if (result.errors) {
        return c.json(errorMessages([result.errors.toString()], result.errors), 500);
      }
    }
    if (pretty) {
      const payload = JSON.stringify(result, null, pretty ? 2 : 0);
      return c.text(payload, 200, {
        "Content-Type": "application/json"
      });
    } else {
      return c.json(result);
    }
  };
};
var getGraphQLParams = async (request) => {
  const urlData = new URLSearchParams(request.url.split("?")[1]);
  const bodyData = await parseBody(request);
  let query = urlData.get("query") ?? bodyData.query;
  if (typeof query !== "string") {
    query = null;
  }
  let variables = urlData.get("variables") ?? bodyData.variables;
  if (typeof variables === "string") {
    try {
      variables = JSON.parse(variables);
    } catch {
      throw Error("Variables are invalid JSON.");
    }
  } else if (typeof variables !== "object") {
    variables = null;
  }
  let operationName = urlData.get("operationName") ?? bodyData.operationName;
  if (typeof operationName !== "string") {
    operationName = null;
  }
  const raw = urlData.get("raw") != null || bodyData.raw !== void 0;
  const params = {
    query,
    variables,
    operationName,
    raw
  };
  return params;
};
var errorMessages = (messages, graphqlErrors) => {
  if (graphqlErrors) {
    return {
      errors: graphqlErrors
    };
  }
  return {
    errors: messages.map((message) => {
      return {
        message
      };
    })
  };
};
var respondWithGraphiQL = (c) => {
  return c.html(`<!--
 *  Copyright (c) 2021 GraphQL Contributors
 *  All rights reserved.
 *
 *  This source code is licensed under the license found in the
 *  LICENSE file in the root directory of this source tree.
-->
<!doctype html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>GraphiQL</title>
    <style>
      body {
        margin: 0;
        overflow: hidden; /* in Firefox */
      }

      #graphiql {
        height: 100dvh;
      }

      .loading {
        height: 100%;
        display: flex;
        align-items: center;
        justify-content: center;
        font-size: 4rem;
      }
    </style>
    <link
      rel="stylesheet"
      href="https://esm.sh/graphiql@4.0.0/dist/style.css"
    />
    <link
      rel="stylesheet"
      href="https://esm.sh/@graphiql/plugin-explorer@4.0.0/dist/style.css"
    />
    <!-- Note: the ?standalone flag bundles the module along with all of its \`dependencies\`, excluding \`peerDependencies\`, into a single JavaScript file. -->
    <script type="importmap">
      {
        "imports": {
          "react": "https://esm.sh/react@19.1.0",
          "react/jsx-runtime": "https://esm.sh/react@19.1.0/jsx-runtime",

          "react-dom": "https://esm.sh/react-dom@19.1.0",
          "react-dom/client": "https://esm.sh/react-dom@19.1.0/client",

          "graphiql": "https://esm.sh/graphiql@4.0.0?standalone&external=react,react/jsx-runtime,react-dom,@graphiql/react",
          "@graphiql/plugin-explorer": "https://esm.sh/@graphiql/plugin-explorer@4.0.0?standalone&external=react,react/jsx-runtime,react-dom,@graphiql/react,graphql",
          "@graphiql/react": "https://esm.sh/@graphiql/react@0.30.0?standalone&external=react,react/jsx-runtime,react-dom,graphql,@graphiql/toolkit",

          "@graphiql/toolkit": "https://esm.sh/@graphiql/toolkit@0.11.2?standalone&external=graphql",
          "graphql": "https://esm.sh/graphql@16.11.0"
        }
      }
    </script>
    <script type="module">
      // Import React and ReactDOM
      import React from 'react';
      import ReactDOM from 'react-dom/client';
      // Import GraphiQL and the Explorer plugin
      import { GraphiQL } from 'graphiql';
      import { createGraphiQLFetcher } from '@graphiql/toolkit';
      import { explorerPlugin } from '@graphiql/plugin-explorer';

      const fetcher = createGraphiQLFetcher({
        url: '${c.req.path}',
      });
      const explorer = explorerPlugin();

      function App() {
        return React.createElement(GraphiQL, {
          fetcher,
          plugins: [explorer],
        });
      }

      const container = document.getElementById('graphiql');
      const root = ReactDOM.createRoot(container);
      root.render(React.createElement(App));
    </script>
  </head>
  <body>
    <div id="graphiql">
      <div class="loading">Loading\u2026</div>
    </div>
  </body>
</html>
`);
};
export {
  errorMessages,
  getGraphQLParams,
  graphqlServer,
  respondWithGraphiQL
};
