// attio/lib/commands/log/subscribe-to-logs.js
import { complete, errored, isErrored } from "@attio/fetchable-npm";
import { Realtime } from "../../util/realtime.js";
import { logEventSchema } from "./log-event.js";
import { LogsBuffer } from "./logs-buffer.js";
export async function subscribeToLogs({ workspaceId, appId, }, listener, onConnectionSuspended) {
    const logsBuffer = new LogsBuffer(200);
    logsBuffer.listen((event) => event.forEach(listener));
    const realtime = new Realtime();
    realtime.onSuspended(() => {
        onConnectionSuspended({
            close: async () => {
                await realtime.close();
            },
            reconnect: async () => {
                const reconnectResult = await realtime.reconnect();
                if (isErrored(reconnectResult)) {
                    return errored({
                        code: "FAILED_TO_CONNECT_TO_LOGS_SERVER",
                        error: reconnectResult.error.error,
                    });
                }
                return complete(true);
            },
        });
    });
    const channelName = `ecosystem-app-logs:${workspaceId}:${appId}`;
    const realtimeSubscriptionResult = await realtime.subscribe(channelName, "app-dev-log-emitted", logEventSchema, (event) => {
        logsBuffer.add(event);
    });
    if (isErrored(realtimeSubscriptionResult)) {
        return errored({
            code: "FAILED_TO_CONNECT_TO_LOGS_SERVER",
            error: realtimeSubscriptionResult.error.error,
        });
    }
    const realtimeSubscription = realtimeSubscriptionResult.value;
    return complete({
        unsubscribe: async () => {
            logsBuffer.close();
            await realtimeSubscription.unsubscribe();
        },
    });
}
