// attio/lib/api/api.js
import { complete, errored, isErrored } from "@attio/fetchable-npm";
import { APP } from "../env.js";
import { SDK_VERSION } from "../sdk-version.js";
import { Fetcher } from "./fetcher.js";
import { ablyAuthResponseSchema, appInfoSchema, completeBundleUploadSchema, createDevVersionSchema, createVersionSchema, installationSchema, listDevWorkspacesResponseSchema, startUploadSchema, TEST_APP_INFO, TEST_WORKSPACES, tokenResponseSchema, versionsSchema, whoamiSchema, } from "./schemas.js";
class ApiImpl {
    _fetcher;
    constructor() {
        this._fetcher = new Fetcher();
    }
    async whoami() {
        const result = await this._fetcher.get({
            path: `${APP}/api/auth/whoami`,
            schema: whoamiSchema,
        });
        if (isErrored(result)) {
            return errored({ code: "WHOAMI_ERROR", error: result.error });
        }
        return complete(result.value.user);
    }
    async fetchWorkspaces() {
        if (process.env.NODE_ENV === "test") {
            return complete(TEST_WORKSPACES);
        }
        const result = await this._fetcher.get({
            path: "/dev-workspaces",
            schema: listDevWorkspacesResponseSchema,
        });
        if (isErrored(result)) {
            return errored({ code: "FETCH_WORKSPACES_ERROR", error: result.error });
        }
        return complete(result.value.workspaces);
    }
    async createVersion({ appId, major, cliVersion, }) {
        const result = await this._fetcher.post({
            path: `/apps/${appId}/prod-versions`,
            body: {
                major,
                cli_version: cliVersion,
                sdk_version: SDK_VERSION,
            },
            schema: createVersionSchema,
        });
        if (isErrored(result)) {
            return errored({ code: "CREATE_VERSION_ERROR", error: result.error });
        }
        return result;
    }
    async fetchAppInfo(appSlug) {
        if (process.env.NODE_ENV === "test") {
            return complete(TEST_APP_INFO.app);
        }
        const result = await this._fetcher.get({
            path: `/apps/by-slug/${appSlug}`,
            schema: appInfoSchema,
        });
        if (isErrored(result)) {
            return errored({ code: "FETCH_APP_INFO_ERROR", error: result.error });
        }
        return complete(result.value.app);
    }
    async completeBundleUpload({ appId, devVersionId, bundleId, }) {
        const result = await this._fetcher.post({
            path: `/apps/${appId}/dev-versions/${devVersionId}/bundles/${bundleId}/complete`,
            body: {},
            schema: completeBundleUploadSchema,
        });
        if (isErrored(result)) {
            return errored({ code: "COMPLETE_BUNDLE_UPLOAD_ERROR", error: result.error });
        }
        return complete(undefined);
    }
    async completeProdBundleUpload({ appId, bundleId, major, minor, }) {
        const result = await this._fetcher.post({
            path: `/apps/${appId}/prod-versions/${major}/${minor}/bundles/${bundleId}/complete`,
            body: {},
            schema: completeBundleUploadSchema,
        });
        if (isErrored(result)) {
            return errored({ code: "COMPLETE_PROD_BUNDLE_UPLOAD_ERROR", error: result.error });
        }
        return complete(undefined);
    }
    async startUpload({ appId, devVersionId, }) {
        const result = await this._fetcher.post({
            path: `/apps/${appId}/dev-versions/${devVersionId}/bundles`,
            body: {
                sdk_version: SDK_VERSION,
            },
            schema: startUploadSchema,
        });
        if (isErrored(result)) {
            return errored({ code: "START_UPLOAD_ERROR", error: result.error });
        }
        return result;
    }
    async createDevVersion({ appId, cliVersion, targetWorkspaceId, environmentVariables, }) {
        const result = await this._fetcher.post({
            path: `/apps/${appId}/dev-versions`,
            body: {
                major: 1,
                target_workspace_id: targetWorkspaceId,
                environment_variables: environmentVariables,
                cli_version: cliVersion,
            },
            schema: createDevVersionSchema,
        });
        if (isErrored(result)) {
            return errored({ code: "CREATE_DEV_VERSION_ERROR", error: result.error });
        }
        return result;
    }
    async fetchInstallation({ appId, workspaceId, }) {
        const result = await this._fetcher.get({
            path: `/apps/${appId}/workspace/${workspaceId}/dev-installation`,
            schema: installationSchema,
        });
        if (isErrored(result)) {
            if (result.error.code === "HTTP_ERROR" && result.error.status === 404) {
                return complete(null);
            }
            return errored({ code: "FETCH_INSTALLATION_ERROR", error: result.error });
        }
        return result;
    }
    async fetchVersions(appId) {
        const result = await this._fetcher.get({
            path: `/apps/${appId}/prod-versions`,
            schema: versionsSchema,
        });
        if (isErrored(result)) {
            return errored({ code: "FETCH_VERSIONS_ERROR", error: result.error });
        }
        return complete(result.value.app_prod_versions);
    }
    async exchangeToken({ receivedCode, verifierString, redirectUri, clientId, }) {
        const params = new URLSearchParams();
        params.append("grant_type", "authorization_code");
        params.append("code", receivedCode);
        params.append("client_id", clientId);
        params.append("redirect_uri", redirectUri);
        params.append("code_verifier", verifierString);
        const result = await this._fetcher.post({
            path: `${APP}/oidc/token`,
            body: params,
            schema: tokenResponseSchema,
            authenticated: "Not Authenticated",
        });
        if (isErrored(result)) {
            return errored({ code: "GET_TOKEN_ERROR", error: result.error });
        }
        return result;
    }
    async refreshToken({ refreshToken, clientId, }) {
        const params = new URLSearchParams();
        params.append("grant_type", "refresh_token");
        params.append("refresh_token", refreshToken);
        params.append("client_id", clientId);
        const result = await this._fetcher.post({
            path: `${APP}/oidc/token`,
            body: params,
            schema: tokenResponseSchema,
        });
        if (isErrored(result)) {
            return errored({ code: "REFRESH_TOKEN_ERROR", error: result.error });
        }
        return result;
    }
    async authenticateAbly(tokenParams) {
        const result = await this._fetcher.post({
            path: "/integrations/ably/auth",
            body: {
                token_params: tokenParams,
            },
            schema: ablyAuthResponseSchema,
        });
        if (isErrored(result)) {
            const message = "message" in result.error ? result.error.message : result.error.code;
            return errored({ code: "ABLY_AUTHENTICATION_ERROR", message });
        }
        return result;
    }
}
export const api = new ApiImpl();
