// attio/lib/api/schemas.js
import { z } from "zod";
export const whoamiSchema = z.object({
    user: z.object({
        id: z.string(),
        email_address: z.string(),
        name: z.object({
            full: z.string(),
        }),
    }),
});
export const workspaceResponseSchema = z.object({
    workspace_id: z.string().uuid(),
    slug: z.string(),
    name: z.string(),
    logo_url: z.string().nullable(),
});
export const TEST_WORKSPACES = [
    workspaceResponseSchema.parse({
        workspace_id: "a85e3bcf-a9a2-4df9-9e92-e708bf98d238",
        slug: "test-slug",
        name: "Test Workspace",
        logo_url: null,
    }),
];
export const listDevWorkspacesResponseSchema = z.object({
    workspaces: z.array(workspaceResponseSchema),
});
export const createVersionSchema = z.object({
    app_id: z.string().uuid(),
    major: z.number(),
    minor: z.number(),
    app_prod_version_bundle_id: z.string(),
    client_bundle_upload_url: z.string(),
    server_bundle_upload_url: z.string(),
});
export const appInfoSchema = z.object({
    app: z.object({
        app_id: z.string().uuid(),
        title: z.string(),
    }),
});
export const TEST_APP_INFO = appInfoSchema.parse({
    app: {
        app_id: "0cbafb09-2ef7-473c-a048-08a5d190a395",
        title: "Test App",
    },
});
export const completeBundleUploadSchema = z.object({
    success: z.literal(true),
});
export const startUploadSchema = z.object({
    app_dev_version_id: z.string().uuid(),
    app_dev_version_bundle_id: z.string().uuid(),
    client_bundle_upload_url: z.string().url(),
    server_bundle_upload_url: z.string().url(),
});
export const createDevVersionSchema = z.object({
    app_id: z.string().uuid(),
    app_dev_version_id: z.string(),
});
export const installationSchema = z.object({
    app_id: z.string().uuid(),
    workspace_id: z.string(),
});
export const versionSchema = z.object({
    major: z.number(),
    minor: z.number(),
    created_at: z.string(),
    released_at: z.string().nullable().optional(),
    is_published: z.boolean(),
    num_installations: z.number(),
    publication_status: z.enum(["private", "in-review", "published", "rejected"]),
});
export const versionsSchema = z.object({
    app_prod_versions: z.array(versionSchema),
});
export const tokenResponseSchema = z.object({
    access_token: z.string(),
    token_type: z.literal("Bearer"),
    refresh_token: z.string(),
    expires_in: z.number(),
});
export const ablyAuthResponseSchema = z.object({
    capability: z.string(),
    clientId: z.string().optional(),
    expires: z.number(),
    issued: z.number(),
    token: z.string(),
});
