// attio/lib/client/forms/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.string = exports.plainDate = exports.number = exports.dateTime = exports.boolean = exports.attioRecord = exports.array = void 0;
var array_1 = require("./array");
Object.defineProperty(exports, "array", { enumerable: true, get: function () { return array_1.array; } });
var attio_record_1 = require("./attio-record");
Object.defineProperty(exports, "attioRecord", { enumerable: true, get: function () { return attio_record_1.attioRecord; } });
var boolean_1 = require("./boolean");
Object.defineProperty(exports, "boolean", { enumerable: true, get: function () { return boolean_1.boolean; } });
var date_time_1 = require("./date-time");
Object.defineProperty(exports, "dateTime", { enumerable: true, get: function () { return date_time_1.dateTime; } });
var number_1 = require("./number");
Object.defineProperty(exports, "number", { enumerable: true, get: function () { return number_1.number; } });
var plain_date_1 = require("./plain-date");
Object.defineProperty(exports, "plainDate", { enumerable: true, get: function () { return plain_date_1.plainDate; } });
var string_1 = require("./string");
Object.defineProperty(exports, "string", { enumerable: true, get: function () { return string_1.string; } });
