// attio/lib/util/find-surface-exports/surface-types.js
export function getSurfaceTypeName(name) {
    return name.charAt(0).toUpperCase() + name.slice(1);
}
const recordActionTypeDefinition = `
export interface ${getSurfaceTypeName("recordAction")} {
    id: string
    label: string
    icon: string
    onTrigger: (record: {recordId: string; object: string}) => void | Promise<void>
    objects?: ObjectSlug | Array<ObjectSlug> | ((attributes: Array<Attribute>) => boolean)
}
`;
const bulkRecordActionTypeDefinition = `
export interface ${getSurfaceTypeName("bulkRecordAction")} {
    id: string
    label: string
    icon: string
    onTrigger: (record: {recordIds: Array<string>; object: string}) => void | Promise<void>
    objects?: ObjectSlug | Array<ObjectSlug> | ((attributes: Array<Attribute>) => boolean)
}
`;
const recordWidgetTypeDefinition = `
export interface ${getSurfaceTypeName("recordWidget")} {
    id: string
    Widget: (props: {recordId: string; object: ObjectSlug}) => React.JSX.Element
    objects?: ObjectSlug | Array<ObjectSlug> | ((attributes: Array<Attribute>) => boolean)
}
`;
const callRecordingInsightTextSelectionActionTypeDefinition = `
export interface ${getSurfaceTypeName("callRecordingInsightTextSelectionAction")} {
    id: string
    label: string
    icon: string
    onTrigger: (selection: {text: string; markdown: string}) => void | Promise<void>
}
`;
const callRecordingSummaryTextSelectionActionTypeDefinition = `
export interface ${getSurfaceTypeName("callRecordingSummaryTextSelectionAction")} {
    id: string
    label: string
    icon: string
    onTrigger: (selection: {text: string; markdown: string}) => void | Promise<void>
}
`;
const callRecordingTranscriptTextSelectionActionTypeDefinition = `
export interface ${getSurfaceTypeName("callRecordingTranscriptTextSelectionAction")} {
    id: string
    label: string
    icon: string
    onTrigger: (selection: {transcript: Array<{speaker: string; text: string}>; url: string}) => void | Promise<void>
}
`;
const workflowBlockTypeDefinition = `
export interface ${getSurfaceTypeName("workflowBlock")} {
    id: string
    title: string
    configSchema: Record<string, Node> | Struct
    configurator: React.ReactNode
}
`;
const workspaceSettingsTypeDefinition = `
export interface ${getSurfaceTypeName("workspaceSettings")} {
    Page: React.ComponentType<{}>
}
`;
export const surfaceTypeDefinitions = `
${recordActionTypeDefinition}

${bulkRecordActionTypeDefinition}

${recordWidgetTypeDefinition}

${callRecordingInsightTextSelectionActionTypeDefinition}

${callRecordingSummaryTextSelectionActionTypeDefinition}

${callRecordingTranscriptTextSelectionActionTypeDefinition}

${workspaceSettingsTypeDefinition}

${workflowBlockTypeDefinition}
`;
